src/ctia/entity/vulnerability/cpe.clj

```diff
@@ -0,0 +1,53 @@
+(ns ctia.entity.vulnerability.cpe)
+
+(defn apply-operator
+  [{:keys [negate operands operator result]}]
+  (case operator
+    "OR" (when (if negate
+                 (not (some identity operands))
+                 (some identity operands))
+           result)
+    "AND" (when (if negate
+                  (not (every? identity operands))
+                  (every? identity operands))
+            result)))
+
+(defn query-cpe-matches
+  [{:keys [cpe_match negate operator]} cpes]
+  (apply-operator {:negate negate
+                   :operands (for [uri (map :cpe23Uri cpe_match)]
+                               (some #{uri} cpes))
+                   :operator operator
+                   :result cpe_match}))
+
+(defn query-children
+  [{:keys [children negate operator]} cpes]
+  (apply-operator {:negate negate
+                   :operands (map (fn query-child [child]
+                                    (query-cpe-matches child cpes))
+                                  children)
+                   :operator operator
+                   :result children}))
+
+(defn query-node
+  [{:keys [children cpe_match] :as node} cpes]
+  (cond
+    (seq cpe_match) (query-cpe-matches node cpes)
+    (seq children) (query-children node cpes)))
+
+(defn query-vulnerability
+  [cpes {id :id {:keys [nodes]} :configurations}]
+  (when (seq (keep (fn [node]
+                     (query-node node cpes))
+                   nodes))
+    id))
+
+(defn ids
+  [cpes vulnerabilities]
+  (reduce (fn ids-rf
```

Comment:
If I am not mistaken this reduce is a [`keep`](https://clojuredocs.org/clojure.core/keep), isn't it?

Reply:
Yes it is - thank you.